src/System.Memory/src/System/Buffers/ReadOnlyBufferExtensions.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#pragma warning disable 1591
+
+namespace System.Buffers
+{
+    public static class ReadOnlyBufferExtensions
+    {
+        public static SequencePosition? PositionOf<T>(this ReadOnlyBuffer<T> buffer, T value) where T: IEquatable<T>
+        {
+            SequencePosition position = buffer.Start;
+            SequencePosition result = position;
+            while (buffer.TryGet(ref position, out var memory))
+            {
+                var index = memory.Span.IndexOf(value);
+                if (index != -1)
+                {
+                    return buffer.GetPosition(result, index);
+                }
+                result = position;
+            }
+            return null;
+        }
+
+
+        /// <summary>
+        /// Copy the <see cref="ReadOnlyBuffer{T}"/> to the specified <see cref="Span{Byte}"/>.
+        /// </summary>
+        /// <param name="buffer">The source <see cref="ReadOnlyBuffer{T}"/>.</param>
+        /// <param name="destination">The destination <see cref="Span{Byte}"/>.</param>
+        public static void CopyTo<T>(this ReadOnlyBuffer<T> buffer, Span<T> destination)
+        {
+            if (buffer.Length > destination.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.destination);
+            }
+
+            foreach (var segment in buffer)
+            {
+                segment.Span.CopyTo(destination);
+                destination = destination.Slice(segment.Length);
+            }
+        }
+
+        /// <summary>
+        /// Converts the <see cref="ReadOnlyBuffer{T}"/> to an array/>
+        /// </summary>
+        public static T[] ToArray<T>(this ReadOnlyBuffer<T> buffer)
+        {
+            var array = new T[buffer.Length];
```

Comment:
buffer.Length is a long. Do we need checks for it being too large?

Reply:
How do you define too large?

Reply:
The array new operator does these checks for you.

Reply:
> The array new operator does these checks for you.

I see. Thanks for clarifying.

Reply:
> The array new operator does these checks for you.

@jkotas, can you point me to the source code for this please?

Reply:
It is done by Roslyn. For example, if you compile this:

```csharp
static byte[] foo(long l) => new byte[l];
```

It produces IL like this:

```
    IL_0000:  ldarg.0
    IL_0001:  conv.ovf.i <- this does the long->int conversion with overflow check
    IL_0002:  newarr     [mscorlib]System.Byte
    IL_0007:  ret
```